### PR TITLE
check if the token is a lifetime before parsing

### LINKED ISSUE #50381: ICE: non-lifetime passed to :lifetime
This code crashes the compiler. However, the location of the error is pointed out, which is nice.

```rust
#![feature(macro_lifetime_matcher)]

macro_rules! m { ($x:lifetime) => { } }

m!(a);
```

```
error: internal compiler error: not a lifetime
 --> src/main.rs:5:8
  |
5 |     m!(a);
  |        ^

thread 'main' panicked at 'Box<Any>', librustc_errors/lib.rs:491:9
note: Run with `RUST_BACKTRACE=1` for a backtrace.
error: aborting due to previous error
```

cc #50380 @ricochet1k

### PR DESCRIPTION
**Pull request by rleungx**
Fixes #50381.

### CODE REVIEW

File: src/test/compile-fail/macro-non-lifetime.rs [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+// Test for issue #50381: non-lifetime passed to :lifetime.
+
+#![feature(macro_lifetime_matcher)]
+
+macro_rules! m { ($x:lifetime) => { } }
+
+fn main() {
+    m!(a);
+    //~^ ERROR expected a lifetime, found `a`
+}
```
Comment (estebank) [inline src/test/compile-fail/macro-non-lifetime.rs]: In the future, add tests to the `src/test/ui/` directory instead. That way the test suite will also check against the human output, not only that the error message is present.

**Review by estebank — APPROVED** [review]
@bors r+ rollup

File: src/test/compile-fail/macro-non-lifetime.rs (hunk 1 above)
Comment (rleungx) [reply src/test/compile-fail/macro-non-lifetime.rs]: Got it

**Pull request merged**